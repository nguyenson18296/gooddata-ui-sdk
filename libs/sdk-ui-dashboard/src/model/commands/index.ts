// (C) 2021 GoodData Corporation

import { LoadDashboard, RenameDashboard, ResetDashboard, SaveDashboard, SaveDashboardAs } from "./dashboard";
import {
    AddAttributeFilter,
    ChangeAttributeFilterSelection,
    ChangeDateFilterSelection,
    MoveAttributeFilter,
    RemoveAttributeFilters,
    SetAttributeFilterParent,
} from "./filters";
import {
    ChangeInsightWidgetFilterSettings,
    ChangeInsightWidgetHeader,
    ChangeInsightWidgetInsight,
    ChangeInsightWidgetVisProperties,
    ModifyDrillsForInsightWidget,
    RefreshInsightWidget,
    RemoveDrillsForInsightWidget,
} from "./insight";
import {
    ChangeKpiWidgetComparison,
    ChangeKpiWidgetFilterSettings,
    ChangeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    RefreshKpiWidget,
} from "./kpi";
import {
    AddLayoutSection,
    AddSectionItems,
    ChangeLayoutSectionHeader,
    MoveLayoutSection,
    MoveSectionItem,
    RemoveLayoutSection,
    RemoveSectionItem,
    ReplaceSectionItem,
    UndoLayoutChanges,
} from "./layout";
import { CreateAlert, RemoveAlert, UpdateAlert } from "./alerts";
import { CreateScheduledEmail } from "./scheduledEmail";
import {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
} from "./drill";

export { DashboardCommandType, IDashboardCommand, CommandProcessingMeta } from "./base";
export {
    InitialLoadCorrelationId,
    LoadDashboard,
    loadDashboard,
    SaveDashboardAs,
    saveDashboardAs,
    SaveDashboard,
    saveDashboard,
    RenameDashboard,
    renameDashboard,
    ResetDashboard,
    resetDashboard,
} from "./dashboard";
export {
    ChangeDateFilterSelection,
    changeDateFilterSelection,
    clearDateFilterSelection,
    AddAttributeFilter,
    addAttributeFilter,
    MoveAttributeFilter,
    moveAttributeFilter,
    RemoveAttributeFilters,
    removeAttributeFilter,
    ChangeAttributeFilterSelection,
    AttributeFilterSelectionType,
    resetAttributeFilterSelection,
    changeAttributeFilterSelection,
    SetAttributeFilterParent,
    setAttributeFilterParent,
} from "./filters";
export {
    AddLayoutSection,
    addLayoutSection,
    MoveLayoutSection,
    moveLayoutSection,
    RemoveLayoutSection,
    removeLayoutSection,
    ChangeLayoutSectionHeader,
    changeLayoutSectionHeader,
    AddSectionItems,
    addSectionItem,
    ReplaceSectionItem,
    replaceSectionItem,
    MoveSectionItem,
    moveSectionItem,
    RemoveSectionItem,
    eagerRemoveSectionItem,
    UndoLayoutChanges,
    undoLayoutChanges,
    revertLastLayoutChange,
    DashboardLayoutCommands,
    UndoPointSelector,
} from "./layout";
export { CreateAlert, createAlert, RemoveAlert, removeAlert, UpdateAlert, updateAlert } from "./alerts";
export { CreateScheduledEmail } from "./scheduledEmail";
export {
    Drill,
    DrillDown,
    DrillToAttributeUrl,
    DrillToCustomUrl,
    DrillToDashboard,
    DrillToInsight,
    DrillToLegacyDashboard,
    drill,
    drillDown,
    drillToAttributeUrl,
    drillToCustomUrl,
    drillToDashboard,
    drillToInsight,
    drillToLegacyDashboard,
} from "./drill";
export {
    ChangeKpiWidgetHeader,
    changeKpiWidgetHeader,
    ChangeKpiWidgetMeasure,
    changeKpiWidgetMeasure,
    ChangeKpiWidgetFilterSettings,
    changeKpiWidgetFilterSettings,
    ChangeKpiWidgetComparison,
    changeKpiWidgetComparison,
    RefreshKpiWidget,
    refreshKpiWidget,
    KpiWidgetComparison,
} from "./kpi";

export {
    ChangeInsightWidgetHeader,
    changeInsightWidgetHeader,
    ChangeInsightWidgetFilterSettings,
    changeInsightWidgetFilterSettings,
    ChangeInsightWidgetVisProperties,
    changeInsightWidgetVisProperties,
    ChangeInsightWidgetInsight,
    changeInsightWidgetInsight,
    ModifyDrillsForInsightWidget,
    modifyDrillForInsightWidget,
    RemoveDrillsForInsightWidget,
    removeDrillForInsightWidget,
    RemoveDrillsSelector,
    RefreshInsightWidget,
    refreshInsightWidget,
} from "./insight";

/**
 * @internal
 */
export type DashboardCommands =
    | LoadDashboard
    | SaveDashboard
    | SaveDashboardAs
    | RenameDashboard
    | ResetDashboard
    | ChangeDateFilterSelection
    | AddAttributeFilter
    | RemoveAttributeFilters
    | MoveAttributeFilter
    | ChangeAttributeFilterSelection
    | SetAttributeFilterParent
    | AddLayoutSection
    | MoveLayoutSection
    | RemoveLayoutSection
    | ChangeLayoutSectionHeader
    | AddSectionItems
    | ReplaceSectionItem
    | MoveSectionItem
    | RemoveSectionItem
    | UndoLayoutChanges
    | ChangeKpiWidgetHeader
    | ChangeKpiWidgetMeasure
    | ChangeKpiWidgetFilterSettings
    | ChangeKpiWidgetComparison
    | RefreshKpiWidget
    | ChangeInsightWidgetHeader
    | ChangeInsightWidgetFilterSettings
    | ChangeInsightWidgetVisProperties
    | ChangeInsightWidgetInsight
    | ModifyDrillsForInsightWidget
    | RemoveDrillsForInsightWidget
    | RefreshInsightWidget
    | CreateAlert
    | UpdateAlert
    | RemoveAlert
    | CreateScheduledEmail
    | Drill
    | DrillDown
    | DrillToAttributeUrl
    | DrillToCustomUrl
    | DrillToDashboard
    | DrillToInsight
    | DrillToLegacyDashboard;
