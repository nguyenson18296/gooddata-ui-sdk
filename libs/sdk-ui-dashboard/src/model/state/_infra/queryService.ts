// (C) 2021 GoodData Corporation
import { DashboardContext } from "../../types/commonTypes";
import { SagaIterator } from "redux-saga";
import {
    CaseReducer,
    createEntityAdapter,
    createSelector,
    createSlice,
    EntityId,
    EntityState,
    IdSelector,
    OutputSelector,
    PayloadAction,
    Slice,
} from "@reduxjs/toolkit";
import { DashboardQueryType, IDashboardQuery } from "../../queries/base";
import { DashboardState } from "../types";
import { CaseReducerActions } from "@reduxjs/toolkit/src/createSlice";
import { call, put, SagaReturnType, select } from "redux-saga/effects";

/**
 *
 */
export interface IDashboardQueryService<TQuery extends IDashboardQuery, TResult> {
    name: DashboardQueryType;
    generator: (ctx: DashboardContext, query: TQuery) => SagaIterator<TResult>;
    cache: QueryCache<TQuery, TResult>;
}

type QueryCacheEntry<TQuery extends IDashboardQuery, TResult> = {
    query: TQuery;
    result: TResult;
};

type QueryCacheReducer<TQuery extends IDashboardQuery, TResult, TPayload> = CaseReducer<
    EntityState<QueryCacheEntry<TQuery, TResult>>,
    PayloadAction<TPayload>
>;

type AllQueryCacheReducers<TQuery extends IDashboardQuery, TResult> = {
    set: QueryCacheReducer<TQuery, TResult, QueryCacheEntry<TQuery, TResult>>;
    remove: QueryCacheReducer<TQuery, TResult, string>;
    removeAll: QueryCacheReducer<TQuery, TResult, void>;
};

type QueryActions<TQuery extends IDashboardQuery, TResult> = CaseReducerActions<
    AllQueryCacheReducers<TQuery, TResult>
>;

type QueryCache<TQuery extends IDashboardQuery, TResult> = {
    slice: Slice<
        EntityState<QueryCacheEntry<TQuery, TResult>>,
        AllQueryCacheReducers<TQuery, TResult>,
        string
    >;
    actions: QueryActions<TQuery, TResult>;
    selectById: (
        id: EntityId,
    ) => OutputSelector<DashboardState, TResult | undefined, (res: DashboardState) => TResult | undefined>;
};

function createQueryCacheSlice<TQuery extends IDashboardQuery, TResult>(
    queryName: string,
    selectId: IdSelector<TQuery>,
): QueryCache<TQuery, TResult> {
    const cacheEntryId = (entry: QueryCacheEntry<TQuery, TResult>) => selectId(entry.query);
    const adapter = createEntityAdapter({ selectId: cacheEntryId });
    // TODO: getting massive typing error here, don't know how to solve it
    const reducers: AllQueryCacheReducers<TQuery, TResult> = {
        set: adapter.addOne as any,
        remove: adapter.removeOne as any,
        removeAll: adapter.removeAll as any,
    };

    const slice = createSlice<
        EntityState<QueryCacheEntry<TQuery, TResult>>,
        AllQueryCacheReducers<TQuery, TResult>
    >({
        name: queryName,
        initialState: adapter.getInitialState(),
        reducers,
    });

    const selectors = adapter.getSelectors((state: DashboardState) => state._queryCache[queryName]);
    const selectById = (id: EntityId) => {
        return createSelector(
            (state: DashboardState) => state,
            (state) => {
                return selectors.selectById(state, id)?.result;
            },
        );
    };

    return {
        slice,
        actions: slice.actions,
        selectById,
    };
}

export type QueryServiceCachingConfig<TQuery extends IDashboardQuery, TResult> = {
    queryName: DashboardQueryType;
    generator: (ctx: DashboardContext, query: TQuery) => SagaIterator<TResult>;
    queryToCacheKey: (query: TQuery) => EntityId;
};

export function createCachedQueryService<TQuery extends IDashboardQuery, TResult>(
    config: QueryServiceCachingConfig<TQuery, TResult>,
): IDashboardQueryService<TQuery, TResult> {
    const { queryName, queryToCacheKey, generator } = config;
    const queryCache = createQueryCacheSlice<TQuery, TResult>(queryName, queryToCacheKey);

    function* generatorWithQueryCache(ctx: DashboardContext, query: TQuery): SagaIterator<TResult> {
        const cacheKey = queryToCacheKey(query);
        const cachedResult = yield select(queryCache.selectById(cacheKey));

        if (cachedResult) {
            return cachedResult;
        }

        const result: SagaReturnType<typeof generator> = yield call(generator, ctx, query);

        yield put(
            queryCache.actions.set({
                query,
                result,
            }),
        );

        return result;
    }

    return {
        name: queryName,
        generator: generatorWithQueryCache,
        cache: queryCache,
    };
}
