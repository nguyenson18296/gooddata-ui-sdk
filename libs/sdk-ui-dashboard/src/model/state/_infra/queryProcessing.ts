// (C) 2021 GoodData Corporation

import { IDashboardQueryService } from "./queryService";
import { Saga, SagaIterator } from "redux-saga";
import { actionChannel, call, getContext, spawn, take } from "redux-saga/effects";
import { IDashboardQuery } from "../../queries";
import { DashboardContext } from "../../types/commonTypes";
import keyBy from "lodash/keyBy";
import { Action, CombinedState, combineReducers, Reducer } from "@reduxjs/toolkit";
import fromPairs from "lodash/fromPairs";
import noop from "lodash/noop";

/**
 * Query processing component has multiple pieces that need to be integrated into the redux store.
 */
export type QueryProcessingComponents = {
    /**
     * Query services may store the results in state for caching purposes. All services that use caching implement
     * the cache as a separate slice of the internal `_queryCache` part of the state. This reducer is a combined
     * reducer including all the appropriate slice reducers.
     */
    queryCacheReducer: Reducer<CombinedState<any>>;

    /**
     * A single saga is in place to handle query processing requests. Query requests will be processes concurrently.
     */
    rootQueryProcessor: Saga;
};

export const QueryEnvelopeActionTypeName = "@@QUERY.ENVELOPE";

export type QueryEnvelope = {
    readonly type: typeof QueryEnvelopeActionTypeName;
    readonly query: IDashboardQuery;
    readonly onSuccess: (result: any) => void;
    readonly onError: (err: Error) => void;
};

function* processQuery(
    service: IDashboardQueryService<any, any>,
    ctx: DashboardContext,
    envelope: QueryEnvelope,
) {
    try {
        const result: SagaIterator<typeof service.generator> = yield call(
            service.generator,
            ctx,
            envelope.query,
        );

        envelope.onSuccess(result);
    } catch (e) {
        envelope.onError(e);
    }
}

function ensureQueryWrappedInEnvelope(action: Action): QueryEnvelope {
    if (action.type === QueryEnvelopeActionTypeName) {
        return action as QueryEnvelope;
    }

    return {
        type: QueryEnvelopeActionTypeName,
        query: action,
        onSuccess: noop,
        onError: noop,
    };
}

export function createQueryProcessingComponents(
    queryServices: IDashboardQueryService<any, any>[],
): QueryProcessingComponents {
    const servicesByType = keyBy(queryServices, (service) => service.name);
    const queryToReducers = fromPairs(
        queryServices.map((service) => [service.name, service.cache.slice.reducer]),
    );

    return {
        queryCacheReducer: combineReducers(queryToReducers),
        rootQueryProcessor: function* (): SagaIterator<void> {
            const queryChannel = yield actionChannel(
                (action: any) =>
                    action.type === QueryEnvelopeActionTypeName || action.type.startsWith("GDC.DASH/QUERY."),
            );

            while (true) {
                const envelope = ensureQueryWrappedInEnvelope(yield take(queryChannel));
                const dashboardContext: DashboardContext = yield getContext("dashboardContext");
                const service = servicesByType[envelope.query.type];

                yield spawn(processQuery, service, dashboardContext, envelope);
            }
        },
    };
}
