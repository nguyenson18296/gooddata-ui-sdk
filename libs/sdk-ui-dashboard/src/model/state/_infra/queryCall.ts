// (C) 2021 GoodData Corporation

import { IDashboardQuery } from "../../queries";
import noop from "lodash/noop";
import { QueryEnvelope, QueryEnvelopeActionTypeName } from "./queryProcessing";
import { call, put } from "redux-saga/effects";
import { SagaIterator } from "redux-saga";

function envelopeWithPromise<TResult>(q: IDashboardQuery<TResult>): {
    promise: Promise<TResult>;
    envelope: QueryEnvelope;
} {
    const partialEnvelope = {
        onStart: noop,
        onSuccess: noop,
        onError: noop,
    };

    const promise = new Promise<TResult>((resolve, reject) => {
        partialEnvelope.onSuccess = resolve;
        partialEnvelope.onError = reject;
    });

    const envelope: QueryEnvelope = {
        type: QueryEnvelopeActionTypeName,
        query: q,
        ...partialEnvelope,
    };

    return {
        promise,
        envelope,
    };
}

/**
 * Runs the provided query and returns its result.
 *
 * @param q - query to run
 */
export function* query<TResult>(q: IDashboardQuery<TResult>): SagaIterator<TResult> {
    const { promise, envelope } = envelopeWithPromise(q);
    const waitForResult = (): Promise<TResult> => {
        return promise;
    };

    /*
     * dispatch the enveloped query. the envelope is setup so that upon query finish a promise will be resolved
     */
    yield put(envelope);

    return yield call(waitForResult);
}
