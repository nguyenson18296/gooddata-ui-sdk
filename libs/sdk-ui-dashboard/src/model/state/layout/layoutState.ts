// (C) 2021 GoodData Corporation

import { IDashboardLayout } from "@gooddata/sdk-backend-spi";
import { InitialUndoState, UndoEnhancedState } from "../_infra/undoEnhancer";
import { ExtendedDashboardItem, ExtendedDashboardWidget } from "../../types/layoutTypes";
import { DashboardLayoutCommands } from "../../commands";

/**
 * @alpha
 */
export type LayoutStash = Record<string, ExtendedDashboardItem[]>;

/**
 * @alpha
 */
export interface LayoutState extends UndoEnhancedState<DashboardLayoutCommands> {
    layout?: IDashboardLayout<ExtendedDashboardWidget>;
    stash: LayoutStash;
}

export const layoutInitialState: LayoutState = {
    layout: undefined,
    stash: {},
    ...InitialUndoState,
};
