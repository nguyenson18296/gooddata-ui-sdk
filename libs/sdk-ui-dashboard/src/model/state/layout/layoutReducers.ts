// (C) 2021 GoodData Corporation
import { CaseReducer, PayloadAction } from "@reduxjs/toolkit";
import { LayoutState } from "./layoutState";
import {
    IDashboardLayout,
    IDashboardLayoutSectionHeader,
    isInsightWidget,
    isKpiWidget,
} from "@gooddata/sdk-backend-spi";
import { invariant } from "ts-invariant";
import { undoReducer, withUndo } from "../_infra/undoEnhancer";
import {
    ExtendedDashboardItem,
    ExtendedDashboardLayoutSection,
    RelativeIndex,
    StashedDashboardItemsId,
} from "../../types/layoutTypes";
import { addArrayElements, moveArrayElement, removeArrayElement } from "../../utils/arrayOps";
import { areObjRefsEqual, ObjRef } from "@gooddata/sdk-model";
import { WidgetHeader } from "../../types/widgetTypes";
import flatMap from "lodash/flatMap";

type LayoutReducer<A> = CaseReducer<LayoutState, PayloadAction<A>>;

//
//
//

const setLayout: LayoutReducer<IDashboardLayout> = (state, action) => {
    state.layout = action.payload;
};

//
//
//

type AddSectionActionPayload = {
    section: ExtendedDashboardLayoutSection;
    index: RelativeIndex;
    usedStashes: StashedDashboardItemsId[];
};

const addSection: LayoutReducer<AddSectionActionPayload> = (state, action) => {
    invariant(state.layout);

    const { index, section, usedStashes } = action.payload;

    addArrayElements(state.layout.sections, index, [section]);

    usedStashes.forEach((stashIdentifier) => {
        delete state.stash[stashIdentifier];
    });
};

//
//
//

type RemoveSectionActionPayload = { index: RelativeIndex; stashIdentifier?: StashedDashboardItemsId };

const removeSection: LayoutReducer<RemoveSectionActionPayload> = (state, action) => {
    invariant(state.layout);

    const { index, stashIdentifier } = action.payload;

    if (stashIdentifier) {
        const items = state.layout.sections[index].items;

        state.stash[stashIdentifier] = items;
    }

    removeArrayElement(state.layout.sections, index);
};

//
//
//

type ChangeSectionActionPayload = { index: number; header: IDashboardLayoutSectionHeader };

const changeSectionHeader: LayoutReducer<ChangeSectionActionPayload> = (state, action) => {
    invariant(state.layout);

    const { index, header } = action.payload;

    state.layout.sections[index].header = header;
};

//
//
//

type MoveSectionActionPayload = { sectionIndex: number; toIndex: RelativeIndex };

const moveSection: LayoutReducer<MoveSectionActionPayload> = (state, action) => {
    invariant(state.layout);

    const { sectionIndex, toIndex } = action.payload;

    moveArrayElement(state.layout.sections, sectionIndex, toIndex);
};

//
//
//

type AddSectionItemsActionPayload = {
    sectionIndex: number;
    itemIndex: number;
    items: ExtendedDashboardItem[];
    usedStashes: StashedDashboardItemsId[];
};

const addSectionItems: LayoutReducer<AddSectionItemsActionPayload> = (state, action) => {
    invariant(state.layout);

    const { sectionIndex, itemIndex, items, usedStashes } = action.payload;
    const section = state.layout.sections[sectionIndex];

    invariant(section);

    addArrayElements(section.items, itemIndex, items);

    usedStashes.forEach((stashIdentifier) => {
        delete state.stash[stashIdentifier];
    });
};

//
//
//

type MoveSectionItemActionPayload = {
    sectionIndex: number;
    itemIndex: number;
    toSectionIndex: number;
    toItemIndex: RelativeIndex;
};

const moveSectionItem: LayoutReducer<MoveSectionItemActionPayload> = (state, action) => {
    invariant(state.layout);

    const { sectionIndex, itemIndex, toSectionIndex, toItemIndex } = action.payload;
    const fromSection = state.layout.sections[sectionIndex];
    const toSection = state.layout.sections[toSectionIndex];

    invariant(fromSection);
    invariant(toSection);

    const item = removeArrayElement(fromSection.items, itemIndex);

    invariant(item);

    addArrayElements(toSection.items, toItemIndex, [item]);
};

//
//
//

type RemoveSectionItemActionPayload = {
    sectionIndex: number;
    itemIndex: number;
    stashIdentifier?: StashedDashboardItemsId;
};

const removeSectionItem: LayoutReducer<RemoveSectionItemActionPayload> = (state, action) => {
    invariant(state.layout);

    const { sectionIndex, itemIndex, stashIdentifier } = action.payload;
    const section = state.layout.sections[sectionIndex];

    invariant(section);

    const item = removeArrayElement(section.items, itemIndex);

    invariant(item);

    if (stashIdentifier) {
        state.stash[stashIdentifier] = [item];
    }
};

//
//
//

type ReplaceSectionItemActionPayload = {
    sectionIndex: number;
    itemIndex: number;
    newItems: ExtendedDashboardItem[];
    stashIdentifier?: StashedDashboardItemsId;
    usedStashes: StashedDashboardItemsId[];
};

const replaceSectionItem: LayoutReducer<ReplaceSectionItemActionPayload> = (state, action) => {
    invariant(state.layout);

    const { sectionIndex, itemIndex, newItems, stashIdentifier, usedStashes } = action.payload;
    const section = state.layout.sections[sectionIndex];

    invariant(section);

    const item = removeArrayElement(section.items, itemIndex);

    invariant(item);

    if (stashIdentifier) {
        state.stash[stashIdentifier] = [item];
    }

    addArrayElements(section.items, itemIndex, newItems);

    usedStashes.forEach((usedStash) => {
        /*
         * It is a valid case that the new item is taken from a stash and the replaced item is then
         * used to replace the same stash.
         */
        if (stashIdentifier !== undefined && usedStash === stashIdentifier) {
            return;
        }

        delete state.stash[usedStash];
    });
};

//
//
//

type ReplaceWidgetHeader = {
    ref: ObjRef;
    header: WidgetHeader;
};

const replaceWidgetHeader: LayoutReducer<ReplaceWidgetHeader> = (state, action) => {
    invariant(state.layout);

    const { header, ref: widgetRef } = action.payload;
    const allWidgets = flatMap(state.layout.sections, (section) => section.items.map((item) => item.widget));
    const widget = allWidgets.find((w) => {
        // defer type checks until the actual widget is found
        const ref: ObjRef | undefined = w && (w as any).ref;

        return ref && areObjRefsEqual(ref, widgetRef);
    });

    // this means command handler did not correctly validate that the widget exists before dispatching the
    // reducer action
    invariant(widget && (isKpiWidget(widget) || isInsightWidget(widget)));

    widget.title = header.title ?? "";
};

export const layoutReducers = {
    setLayout,
    addSection: withUndo(addSection),
    removeSection: withUndo(removeSection),
    moveSection: withUndo(moveSection),
    changeSectionHeader: withUndo(changeSectionHeader),
    addSectionItems: withUndo(addSectionItems),
    moveSectionItem: withUndo(moveSectionItem),
    removeSectionItem: withUndo(removeSectionItem),
    replaceSectionItem: withUndo(replaceSectionItem),
    replaceWidgetHeader: withUndo(replaceWidgetHeader),
    undoLayout: undoReducer,
};
