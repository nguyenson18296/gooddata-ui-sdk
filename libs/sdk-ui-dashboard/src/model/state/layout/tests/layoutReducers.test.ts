// (C) 2021 GoodData Corporation
import cloneDeep from "lodash/cloneDeep";
import { SimpleDashboardLayout } from "../../../tests/Dashboard.fixtures";
import { LayoutState } from "../layoutState";
import { InitialUndoState } from "../../_infra/undoEnhancer";
import { IDashboardLayout } from "@gooddata/sdk-backend-spi";
import { layoutReducers } from "../layoutReducers";
import { removeLayoutSection } from "../../../commands";
import { layoutActions } from "../index";
import { produce } from "immer";

describe("layout slice reducer", () => {
    function createLayoutSliceInitialState(layout?: IDashboardLayout): LayoutState {
        const copyOfLayout = cloneDeep(layout);

        return {
            layout: copyOfLayout,
            stash: {},
            ...InitialUndoState,
        };
    }

    it("should correctly handle remove section and create undo entry", () => {
        const initialState = createLayoutSliceInitialState(SimpleDashboardLayout);

        const newState = produce(initialState, (draft) => {
            const removeAction = layoutActions.removeSection({
                index: 0,
                undo: { cmd: removeLayoutSection(0, undefined, "correlation") },
            });

            return layoutReducers.removeSection(draft, removeAction) as any;
        });

        expect(newState.layout!.sections).toEqual([initialState.layout?.sections[1]]);
        expect(newState._undo).toMatchSnapshot();
    });

    it("should correctly stash items when section is removed", () => {
        const initialState = createLayoutSliceInitialState(SimpleDashboardLayout);

        const newState = produce(initialState, (draft) => {
            const removeAction = layoutActions.removeSection({
                index: 0,
                stashIdentifier: "testStash",
                undo: { cmd: removeLayoutSection(0, "testStash", "correlation") },
            });

            return layoutReducers.removeSection(draft, removeAction) as any;
        });

        expect(newState.layout!.sections).toEqual([initialState.layout?.sections[1]]);
        expect(newState.stash["testStash"]).toEqual(initialState.layout!.sections[0].items);
    });
});
