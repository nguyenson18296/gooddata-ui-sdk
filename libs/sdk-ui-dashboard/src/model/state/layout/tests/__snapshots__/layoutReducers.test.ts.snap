// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout slice reducer should correctly handle remove section and create undo entry 1`] = `
Object {
  "undoPointer": 0,
  "undoStack": Array [
    Object {
      "cmd": Object {
        "correlationId": "correlation",
        "payload": Object {
          "index": 0,
          "stashIdentifier": undefined,
        },
        "type": "GDC.DASH/CMD.FLUID_LAYOUT.REMOVE_SECTION",
      },
      "redoPatches": Array [
        Object {
          "op": "replace",
          "path": Array [
            "layout",
            "sections",
            0,
          ],
          "value": Object {
            "items": Array [
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 12,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "description": "",
                  "drills": Array [
                    Object {
                      "origin": Object {
                        "measure": Object {
                          "localIdentifier": "31c22194386b408aa80ab90b966e85a7",
                        },
                        "type": "drillFromMeasure",
                      },
                      "target": Object {
                        "identifier": "aeis6NlXcL7X",
                      },
                      "transition": "in-place",
                      "type": "drillToDashboard",
                    },
                  ],
                  "identifier": "acMpe9e1gOEC",
                  "ignoreDashboardFilters": Array [
                    Object {
                      "displayForm": Object {
                        "uri": "/gdc/md/referenceworkspace/obj/1071",
                      },
                      "type": "attributeFilterReference",
                    },
                  ],
                  "insight": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1308",
                  },
                  "properties": Object {},
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1310",
                  },
                  "title": "Simple Sorted Table",
                  "type": "insight",
                  "uri": "/gdc/md/referenceworkspace/obj/1310",
                },
              },
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 4,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "description": "",
                  "drills": Array [],
                  "identifier": "acrpg7NRgQub",
                  "ignoreDashboardFilters": Array [],
                  "insight": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1308",
                  },
                  "properties": Object {},
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1309",
                  },
                  "title": "Simple Sorted Table",
                  "type": "insight",
                  "uri": "/gdc/md/referenceworkspace/obj/1309",
                },
              },
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 6,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "dateDataSet": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/557",
                  },
                  "description": "",
                  "drills": Array [
                    Object {
                      "origin": Object {
                        "measure": Object {
                          "localIdentifier": "1efdcdfefecb465fb6bee274df905f35",
                        },
                        "type": "drillFromMeasure",
                      },
                      "target": Object {
                        "uri": "/gdc/md/referenceworkspace/obj/1284",
                      },
                      "transition": "pop-up",
                      "type": "drillToInsight",
                    },
                  ],
                  "identifier": "aadaJHsFexBu",
                  "ignoreDashboardFilters": Array [],
                  "insight": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1297",
                  },
                  "properties": Object {},
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1302",
                  },
                  "title": "Test insight",
                  "type": "insight",
                  "uri": "/gdc/md/referenceworkspace/obj/1302",
                },
              },
            ],
            "type": "IDashboardLayoutSection",
          },
        },
        Object {
          "op": "replace",
          "path": Array [
            "layout",
            "sections",
            "length",
          ],
          "value": 1,
        },
      ],
      "undoPatches": Array [
        Object {
          "op": "replace",
          "path": Array [
            "layout",
            "sections",
            0,
          ],
          "value": Object {
            "items": Array [
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 2,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "dateDataSet": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/557",
                  },
                  "description": "",
                  "drills": Array [],
                  "identifier": "abZ3WTf0ckan",
                  "ignoreDashboardFilters": Array [],
                  "kpi": Object {
                    "comparisonDirection": "growIsGood",
                    "comparisonType": "previousPeriod",
                    "metric": Object {
                      "uri": "/gdc/md/referenceworkspace/obj/1271",
                    },
                  },
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1386",
                  },
                  "title": "Amount",
                  "type": "kpi",
                  "uri": "/gdc/md/referenceworkspace/obj/1386",
                },
              },
            ],
            "type": "IDashboardLayoutSection",
          },
        },
        Object {
          "op": "add",
          "path": Array [
            "layout",
            "sections",
            1,
          ],
          "value": Object {
            "items": Array [
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 12,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "description": "",
                  "drills": Array [
                    Object {
                      "origin": Object {
                        "measure": Object {
                          "localIdentifier": "31c22194386b408aa80ab90b966e85a7",
                        },
                        "type": "drillFromMeasure",
                      },
                      "target": Object {
                        "identifier": "aeis6NlXcL7X",
                      },
                      "transition": "in-place",
                      "type": "drillToDashboard",
                    },
                  ],
                  "identifier": "acMpe9e1gOEC",
                  "ignoreDashboardFilters": Array [
                    Object {
                      "displayForm": Object {
                        "uri": "/gdc/md/referenceworkspace/obj/1071",
                      },
                      "type": "attributeFilterReference",
                    },
                  ],
                  "insight": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1308",
                  },
                  "properties": Object {},
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1310",
                  },
                  "title": "Simple Sorted Table",
                  "type": "insight",
                  "uri": "/gdc/md/referenceworkspace/obj/1310",
                },
              },
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 4,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "description": "",
                  "drills": Array [],
                  "identifier": "acrpg7NRgQub",
                  "ignoreDashboardFilters": Array [],
                  "insight": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1308",
                  },
                  "properties": Object {},
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1309",
                  },
                  "title": "Simple Sorted Table",
                  "type": "insight",
                  "uri": "/gdc/md/referenceworkspace/obj/1309",
                },
              },
              Object {
                "size": Object {
                  "xl": Object {
                    "gridWidth": 6,
                  },
                },
                "type": "IDashboardLayoutItem",
                "widget": Object {
                  "dateDataSet": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/557",
                  },
                  "description": "",
                  "drills": Array [
                    Object {
                      "origin": Object {
                        "measure": Object {
                          "localIdentifier": "1efdcdfefecb465fb6bee274df905f35",
                        },
                        "type": "drillFromMeasure",
                      },
                      "target": Object {
                        "uri": "/gdc/md/referenceworkspace/obj/1284",
                      },
                      "transition": "pop-up",
                      "type": "drillToInsight",
                    },
                  ],
                  "identifier": "aadaJHsFexBu",
                  "ignoreDashboardFilters": Array [],
                  "insight": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1297",
                  },
                  "properties": Object {},
                  "ref": Object {
                    "uri": "/gdc/md/referenceworkspace/obj/1302",
                  },
                  "title": "Test insight",
                  "type": "insight",
                  "uri": "/gdc/md/referenceworkspace/obj/1302",
                },
              },
            ],
            "type": "IDashboardLayoutSection",
          },
        },
      ],
    },
  ],
}
`;
